Overhaul EuiDataGrid's cell focus states

chandlerprall opened the pull request:
### Summary

Fixes #2626 

This is ready to review **functionality**, not yet the code side.

Added a new docs page for data grid, **Data grid focus**, to track requirements and provide a datagrid for testing these states. This also touched `EuiFocusTrap`'s outside-click-disabling, so we'll want to double check that (added ability to re-enable a trap which was previously-disabled-by-click).

I also want to diagram/record the focus logic and refactor it, but that has to wait until the logic is solidified and should be in another PR.

### Checklist

~- [ ] Check against **all themes** for compatibility in both light and dark modes~
- [ ] Checked in **mobile**
- [ ] Checked in **IE11** and **Firefox**
~- [ ] Props have proper **autodocs**~
~- [ ] Added **documentation** examples~
- [ ] Added or updated **jest tests**
- [x] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [ ] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

myasonik commented:
When tabbing into the grid for the first time, focus skips all the toolbar items.

## Full steps
1. Start tabbing at the top of the page (tab through "demo", "demo js", and "demo html"

## The bug
Tabbing from "demo html" goes directly to the first cell instead of the first toolbar item

myasonik commented:
OK, I've got a very odd bug...

## Full steps
1. Load page
2. Use mouse to adjust column width
3. Click off the data grid
4. Start tabbing (tab through "demo", "demo js", keep tabbing through all of the toolbar controls)

## THE BUG
After the last toolbar control but before the first cell, notice an extra tab stop where focus is seems to be set on the whole grid

Note: Only tested in Firefox

myasonik commented:
A feature request for the test page (which I otherwise love!) — a toggle on the page to make the columns interactive

chandlerprall commented:
> When tabbing into the grid for the first time, focus skips all the toolbar items.

Fixed, was targeting the wrong div

> After the last toolbar control but before the first cell, notice an extra tab stop where focus is seems to be set on the whole grid

FireFox allows tabbing to a scrollable element. I've added an explicit `tabindex=-1` to disable that.

> A feature request for the test page (which I otherwise love!) — a toggle on the page to make the columns interactive

Not sure I know exactly what you mean; I take this to mean having 3 columns with no interactive elements and the toggle show/hides interactive elements into those?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

myasonik commented:
Confirmed that I can't recreate those bugs anymore! 🎉 

Don't know if this is something you want to handle here or as part of another ticket (as it's not introduced as part of this PR):
If a cell is partially cut off by the scrolling container, there's no way to see the whole thing with just a keyboard. If the cell is completely cut off, the browser will scroll it into view for you.

We should scroll the container to completely show the currently focused cell.

chandlerprall commented:
> If a cell is partially cut off by the scrolling container, there's no way to see the whole thing with just a keyboard. If the cell is completely cut off, the browser will scroll it into view for you.

I noticed this when testing in FF yesterday (Chrome works as expected). Let's deal with that in another issue. Mind creating it?

chandlerprall commented:
Pushed a toggle to enable/disable interactive headers in the focus example, which caught another focus bug.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

myasonik commented:
Not 100% sure what the interaction here should be so just bringing this up for discussion

## Steps
1. Arrow over to a cell with 1 interactive item that is not expandable. 
2. <kbd>tab</kbd> out and back into the grid

## Bug?
Focus is set on the cell, not on the content as it is when arrowing around in the grid. 

(You can also recreate this effect by clicking in the cell but not on the interactive item itself.)

myasonik commented:
## Advanced keyboard control bugs
- Using <kbd>F2</kbd> should be the same as using <kbd>Enter</kbd> when the cell is not expanded (this works as expected) and the same as <kbd>Esc</kbd> when it is expanded (this does not work)
- <kbd>Page Up</kbd> and <kbd>Page Down</kbd> seems to have broken (should set focus to the first/last row in the same column you're currently in

snide commented:
I'm doing a cleanup pass on the docs for this as part of my review. Functionally it seems to make sense. Issues that I see in it such as how to tab through truncated items I think are solvable by judicious use of other bits of functionality in Data Grid such as set widths, expandable cells...etc.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

snide commented:
Sorry, the commit reminded me I still have a todo here. I'll try and get it done this week. I didn't see any major issues functionally. If @myasonik is OK with the state of this PR, I can always PR after since my changes were just to docs.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

chandlerprall commented:
@myasonik all 3 bugs are now fixed!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

chandlerprall commented:
After iterating with @myasonik , <kbd>page up</kbd> & <kbd>page down</kbd> will now always shift focus to the bottom or top (respectively) of the page, not only when actually paging. Also added pagination to the focus example for easier testing.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/

thompsongl commented:
> This is ready to review functionality, not yet the code side.

Let me know when this is ready, @chandlerprall

chandlerprall commented:
@thompsongl this now looks stable for review, thanks!

thompsongl commented on src/components/focus_trap/focus_trap.tsx:
```diff
@@ -57,6 +57,13 @@ export class EuiFocusTrap extends Component<Props, State> {
 
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.disabled === true && this.props.disabled === false) {
+      // eslint-disable-next-line react/no-did-update-set-state
```
Wondering if we should open an issue to convert this to a function component. Avoiding this is probably worth it.

chandlerprall replied:
TBH I've never cared about this react lint rule. We should convert for the sake of converting, at some point. though. Helps avoid these types of bugs

chandlerprall commented:
Will merge on green CI. Thanks everyone!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3128/
